ksql-common/src/main/java/io/confluent/ksql/schema/ksql/KsqlSchema.java

```diff
@@ -40,39 +53,251 @@
+    for (final Field field : getSchema().fields()) {
+      String fieldName = field.name();
+      fieldName = fieldName.substring(fieldName.indexOf(SchemaUtil.FIELD_NAME_DELIMITER) + 1);
+      if (!fieldName.equalsIgnoreCase(SchemaUtil.ROWTIME_NAME)
+          && !fieldName.equalsIgnoreCase(SchemaUtil.ROWKEY_NAME)) {
+        schemaBuilder.field(fieldName, field.schema());
+      }
+    }
+    return KsqlSchema.of(schemaBuilder.build());
+  }
+
+  public Set<Integer> implicitColumnIndexes() {
+    final Set<Integer> indexSet = new HashSet<>();
+    for (int i = 0; i < ((Schema) schema).fields().size(); i++) {
+      final Field field = ((Schema) schema).fields().get(i);
+      if (field.name().equalsIgnoreCase(SchemaUtil.ROWTIME_NAME)
+          || field.name().equalsIgnoreCase(SchemaUtil.ROWKEY_NAME)) {
+        indexSet.add(i);
+      }
+    }
+    return indexSet;
+  }
+
+  public Optional<Field> findField(final String fieldName) {
+    return schema.fields()
+        .stream()
+        .filter(f -> SchemaUtil.matchFieldName(f, fieldName))
+        .findFirst();
+  }
+
+  public OptionalInt findFieldIndex(final String fieldName) {
+    if (schema.fields() == null) {
+      return OptionalInt.empty();
+    }
+
+    for (int i = 0; i < schema.fields().size(); i++) {
+      final Field field = schema.fields().get(i);
+      final int dotIndex = field.name().indexOf(SchemaUtil.FIELD_NAME_DELIMITER);
+      if (dotIndex == -1) {
+        if (field.name().equals(fieldName)) {
+          return OptionalInt.of(i);
+        }
+      } else {
+        if (dotIndex < fieldName.length()) {
+          final String fieldNameWithDot =
+              fieldName.substring(0, dotIndex)
+                  + SchemaUtil.FIELD_NAME_DELIMITER
+                  + fieldName.substring(dotIndex + 1);
+          if (field.name().equals(fieldNameWithDot)) {
+            return OptionalInt.of(i);
+          }
+        }
+      }
+
+    }
+    return OptionalInt.empty();
+  }
+
+  /**
+   * Simpliar to {@link #findFieldIndex(String)}, except requires an exact match.
+   *
+   * @param fieldName the exact name of the field.
+   * @return the name of the field.
+   */
+  public int fieldIndex(final String fieldName) {
+    for (int i = 0; i < schema.fields().size(); i++) {
+      final Field field = schema.fields().get(i);
+      if (field.name().equals(fieldName)) {
+        return i;
+      }
+    }
+
+    throw new KsqlException("Could not find field in schema."
+        + " field: " + fieldName
+        + " schema: " + schema
+    );
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    final KsqlSchema that = (KsqlSchema) o;
+    return schemasAreEqual(schema, that.schema);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(schema);
+  }
+
+  @Override
+  public String toString() {
```

Comment:
we have functionality that relies on this behavior in the QTT, we might want to document that in a comment here

Reply:
I don't believe that's the case. Can you elaborate...

Reply:
Ah, I understand - QTT will fail if someone screws with this, so I don't think it needs a comment